trestle/core/remote/cache.py

```diff
@@ -0,0 +1,487 @@
+        else:
+            raise TrestleError(f'Query failed to run by returning code of ' f'{request.status_code}. {self._query}')
+
+
+class SFTPFetcher(FetcherBase):
+    """Fetcher for SFTP content."""
+
+    # STFP method: https://stackoverflow.com/questions/7563496/open-a-remote-file-using-paramiko-in-python-slow#7563551
+    # For SFTP fetch into memory.
+    def __init__(
+        self,
+        trestle_root: pathlib.Path,
+        uri: str,
+        settings: Settings,
+        refresh: bool = False,
+        fail_hard: bool = False,
+        cache_only: bool = False
+    ) -> None:
+        """Initialize STFP fetcher."""
+        super().__init__(trestle_root, uri, settings, refresh, fail_hard, cache_only)
+        # Is this a valid uri, however? Username and password are optional, of course.
+        u = parse.urlparse(self._uri)
+        if not u.hostname:
+            logger.error(f'Malformed URI, cannot parse hostname in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: missing hostname {self._uri}')
+        if not u.path:
+            logger.error(f'Malformed URI, cannot parse path in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: missing file path {self._uri}')
+        if u.password and not u.username:
+            logger.error(f'Malformed URI, password found but username missing in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: password found but username missing {self._uri}')
+
+        localhost_cached_dir = self._trestle_cache_path / u.hostname
+        # Skip any number of back- or forward slashes preceding the url path (u.path)
+        path_parent = pathlib.Path(u.path[re.search('[^/\\\\]', u.path).span()[0]:]).parent
+        localhost_cached_dir = localhost_cached_dir / path_parent
+        localhost_cached_dir.mkdir(parents=True, exist_ok=True)
+        self._inst_cache_path = localhost_cached_dir / pathlib.Path(pathlib.Path(u.path).name)
+
+    def _sync_cache(self) -> None:
+        u = parse.urlparse(self._uri)
+        client = paramiko.SSHClient()
+
+        if 'SSH_KEY' in os.environ and self._refresh:
+            ssh_key_file = os.environ['SSH_KEY']
+            try:
+                client.load_host_keys(ssh_key_file)
+            except Exception as e:
+                logger.error(f'Error loading host keys from {ssh_key_file}.')
+                logger.debug(e)
+                raise TrestleError(f'Cache update failure for {self._uri}')
+
+        elif 'SSH_KEY' not in os.environ and self._inst_cache_path.parent.exists() and self._refresh:
+            try:
+                client.load_system_host_keys()
+            except Exception as e:
+                logger.error('Error loading system host keys.')
+                logger.debug(e)
+                raise TrestleError(f'Cache update failure for {self._uri}')
+
+        username = getpass.getuser() if not u.username else u.username
```

Comment:
I don't think we want to introduce a command prompt. if there is no username we should default, like normal ssh clients, to the system user.

Reply:
Or we can error.

Reply:
This will not prompt. It just grabs the username. I remember some opinion that this was more portable instead of grabbing os.environ['USER'], but can't say if this is a good opinion, but thought it wouldn't hurt.

Reply:
Ah okay